HW10_interfaces

tetiana-qa7 opened the pull request.

tetiana-qa7 committed:
HW10_interfaces

rodikno commented on HW10/HW10_Interfaces_Abstract_classes/src/Main.java:
```diff
@@ -0,0 +1,67 @@
+        Message fifthMessage = new Message("test");
+        Message six = new Message("new text");
+
+        tania.sendMessage(firstMessage, artem);
```
What if `.sendMessage(Message message, User recipient)` can be modified to retrieve not an actual message but only its content? (the creation of a new Message object can happen inside that method) - this will hide a bit of internal logic so we won't mess with handling messages separately anymore as a user is responsible for creating messages right?

Or we can have even different overloads for `.sendMessage()` method with varying kinds of input params so you can choose which one you want to choose

rodikno commented on HW10/HW10_Interfaces_Abstract_classes/src/Main.java:
```diff
@@ -0,0 +1,67 @@
+        Message six = new Message("new text");
+
+        tania.sendMessage(firstMessage, artem);
+        artem.getMessage(firstMessage);
```
Ideally some potential method `.receiveMessage(Message message)` would handle this situation as usually we don't `receive` messages intentionally - they're being sent to us (it's a passive behavior of a user which can be triggered by somebody when they send the message)

On the other hand, users are reading messages by themselves (this is an active behavior), so it can stay as a direct call from the main method. What if I just want to read the latest message? Ideally I'd just call .readLastMessage() method from the user and it would retrieve the message for me and change its status to "READ" so it can't be read twice

rodikno commented on HW10/HW10_Interfaces_Abstract_classes/src/Main.java:
```diff
@@ -0,0 +1,67 @@
+        tania.getUserContacts();
+        artem.getUserContacts();
+
+        tania.deleteContact("099994444", mum);
```
Having an overloaded version of this method which receives only a phone number OR an object might be also more user-friendly (so we won't need to supply both key and value to delete the contact)

rodikno commented on HW10/HW10_Interfaces_Abstract_classes/src/User.java:
```diff
@@ -0,0 +1,104 @@
+
+    @Override
+    public void sendMessage(Message message, User receiver) {
+        message.setMessageStatus(MessageStatuses.SENT);
```
it makes sense to change the status of the message **after** the rest of operations within the method are done
In this case if any error occurs before the status change - the status won't be set so the message won't stay in the "unwanted state" (imagine if your program preserves the state somehow)

rodikno commented on HW10/HW10_Interfaces_Abstract_classes/src/User.java:
```diff
@@ -0,0 +1,104 @@
+        message.setSender(this);
+        message.setReceiver(receiver);
+        message.getSender();
+        messages.add(message); // щоб поповнити собі арей ліст меседжів новим меседжем
```
Виходить так, що в даному випадку я як відправник якби "зберігаю" цей меседж собі у статусі "відправлене" (оскільки `messages.add(message)` в даному контексті те саме, що `this.messages.add(message)`), але метод приймає об*єкт отримувача, який використовується тільки щоб задати меседжу поле receiver.

Але отримувач не отримає меседж тільки тому, що в самому меседжі він вказаний як "отримувач", вірно?
Тут ми вимушені викликати деякий метод **від імені отримувача**, передавши туди саме повідомлення параметром, щоб отримувач міг зберегти їх у своою колекцію.

rodikno replied:
Підказка `receiver.getMessage(message)`

Псевдокод для метода sendMessage(String messageBody, User recipient)
```java
1. Створити об*єкт повідомлення, задати йому відправника (це поточний об*єкт), отримувача (це об*єкт реципієнта), і контент самого повідомлення (запхнути туди стрінгу, яку було передано параметром `messageBody`)
2. Тепер цей меседж "відправити" реципієнту шляхом виклику метода `retrieveMessage(Message message)` від імені об*єкту реципієнта
3. Змінити статус меседжу на SENT
```

rodikno commented on HW10/HW10_Interfaces_Abstract_classes/src/User.java:
```diff
@@ -0,0 +1,104 @@
+    }
+
+    @Override
+    public void getMessage(Message message) {
```
Не є помилкою, але `getMessage` за іменем можна сплутати з методом-геттером для поля `message`

Так як метод мав би описувати поведінку об*єкта при "отриманні" повідомлення - можливо `retrieveMessage(Message message)` було б більш влучним

Кроки, які по ідеї мав би робити цей метод:
1) Додати отримане повідомлення у власну колекцію повідомлень
2) Змінити статус повідомлення на `RECEIVED`
3) Надрукувати в консоль щось типу "I'm [NAME], received a new unread message from [SENDER_NAME]"

rodikno commented on HW10/HW10_Interfaces_Abstract_classes/src/User.java:
```diff
@@ -0,0 +1,104 @@
+    }
+
+    public void deleteMessage(Message message){
+        messages.removeIf(entry -> "enemy".equals(entry.getMessage()));
```
"enemy"?

rodikno commented on HW10/HW10_Interfaces_Abstract_classes/src/Message.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    @Override
+    public String getReceiver() {
```
`public String getReceiverName()` підійшло б краще оскільки повертається не об*єкт ресівера, а тільки один його атрибут - ім*я

rodikno commented on HW10/HW10_Interfaces_Abstract_classes/src/Message.java:
```diff
@@ -0,0 +1,59 @@
+        System.out.println("New status set successfully");
+    }
+
+    public String toString() {
```
Хороший прийом, але не забувай анонтацію `@Override` пліз, бо це оверрайд методу з класу Object

rodikno commented on HW10/HW10_Interfaces_Abstract_classes/src/User.java:
```diff
@@ -0,0 +1,104 @@
+                "userName='" + name + '}';
+    }
+
+    public void filterMessages(){
```
Good job on adding additional advanced logic!

Although it might be a bit misleading as according to its name `filterMessages` it should filter something (by some criteria) and ideally return a filtered something, right?
It does sorting in place instead and returns a collection which is sorted by status

tetiana-qa7 committed:
added amendments to make the logic better:
- we don't send the Object message any more, when reading, sending message, we can send message using text and identifying receiver
- send message encompasses receiving message by receiver
- we can read the last message
- we can delete contact either by user or by phone and validation
- add logics to check that user is in the contact list before sending

rodikno commented on HW10/HW10_Interfaces_Abstract_classes/src/User.java:
```diff
@@ -31,12 +32,30 @@ public void addContact(String phone, User contact) {
-        contacts.remove(phone, contact);
-        System.out.println("HashMap of contacts WHAT IS LEFT: " + contacts);
-        System.out.println("---");
+    public void deleteContact(String phone) {
```
Прикольний прийом: тепер завдяки перегрузці методу користувач може обрати - видалити контакт по номеру телефону чи явно передаючи контакт об*єктом 👍

rodikno commented on HW10/HW10_Interfaces_Abstract_classes/src/User.java:
```diff
@@ -31,12 +32,30 @@ public void addContact(String phone, User contact) {
+
+
     @Override
     public void getUserContacts() {
```
Некритично, але метод нічого не повертає, тому по назві було б логічніше обрати `printUserContacts()` оскільки слово `get` в назві неявно вказує нам, що метод поверне контакти користувача (колекцією наприклад), а не тільки надрукує їх в консолі

rodikno commented on HW10/HW10_Interfaces_Abstract_classes/src/User.java:
```diff
@@ -54,45 +73,50 @@ public void getUserMessages() {
+            message.setMessageStatus(MessageStatuses.SENT);// для відправемка цей меседж стане відправленим
+            receiver.receiveMessage(message); // пасивна дія, для отримувача цей меседж стане ресівд
+        } else {
+            throw new IllegalArgumentException("You can't send message to someone who is not in your contacts");
```
👍

rodikno commented on HW10/HW10_Interfaces_Abstract_classes/src/User.java:
```diff
@@ -54,45 +73,50 @@ public void getUserMessages() {
-        messages.removeIf(entry -> "enemy".equals(entry.getMessage()));
+    public void readLastMessage() {
+        Message message = messages.get(messages.size() - 1); // отримуємо останнє повідомлення
+        if (!message.getMessageStatus().equals(MessageStatuses.READ)) { // перевірити що вже не прочитано
```
Не хочу докопуватися, але тут приховано логічний баг: а що якщо останній меседж в колекції дійсно виявився вже прочитаним?
Виходить, що хтось ніколи не зможе прочитати останнє непрочитане повідомлення тільки з тієї причини, що колись він вже прочитав останнє зі списку (де їх може бути більше одного.

В ідеалі тут було б непогано спочатку фільтрувати список: тобто зробити так, щоб нижче ти працювала вже тільки з непрочитаними повідомленнями.

Якщо виявляється, що після фільтрації є хоча б одне непрочитане повідомлення - читаємо останнє зі списку і одразу міняємо його статус на "прочитане"
Якщо після фільтрації у нас не виявилось непрочитаних повідомлень - було б вигідніше не викидати ексепшн а просто:
а) повернути null
АБО
б) повернути пусте повідомлення
АБО (оскільки у нас void метод
в) Надрукувати в консоль щось типу "У вас немає непрочитаних повідомлень"

Чому не кидати ексепшн є кращою стратегією? Тому що не мати непрочитаних повідомлень - це стандартна ситуація в контексті бізнес-логіки - а ексепшни ми викидаємо коли щось трапилось не так, вірно? В даному випадку ні юзер, ні сама програма не винні в тому, що непрочитаних повідомлень немає - ну просто ми або вже прочитали всі, або ніхто не прислав нам нових

rodikno commented on HW10/HW10_Interfaces_Abstract_classes/src/User.java:
```diff
@@ -101,4 +125,14 @@ public void filterMessages(){
+
+    public void filterMessages(MessageStatuses status){
+        for (Message element : messages){
+            if(element.getMessageStatus().equals(status)) {
```
Тут тепер дійсно є фільтрація по статусу, але метод друкує результат
тобто краще підійде назва `printFilteredMessages`
якщо він продовжує називатися `filterMessages` то тоді по ідеї повинен повертати колекцію з відфільтрованих меседжів (тобто тип повернення буде `ArrayList<Message>`) і тоді вже логічною буде назва `getFilteredMessagesBy(MessageStatus status)`

rodikno reviewed (approved):
Апрув, але з нюансами :)
